rfc/system/0000-exteranl-plugin-service.md

```diff
@@ -0,0 +1,270 @@
+    """
+    return {"job_id": job_id, "error_code": error_code, "message": message}
+
+```
+**Ideally**, we should provide a simplified interface to this in flytekit. This would mean that the flytekit plugin can simply use this interface to create a local plugin and a backendplugin.
+
+```python=
+class BackendPluginBase:
+
+    def __init__(self, task_type: str, version: str = "v1"):
+        self._task_type = task_type
+        self._version = version
+
+    @property
+    def task_type(self) -> str:
+        return self._task_type
+
+    @property
+    def version(self) -> str:
+        return self._version
+
+    @abstractmethod
+    async def create(self):
+        pass
+
+    @abstractmethod
+    async def poll(self):
+        pass
+    
+    @abstractmethod
+    async def delete(self):
+        pass
+    
+BackendPluginRegistry.register(BQPlugin())
+```
+
+### (Alternative) GRPC Backend Server
+```python=
+class BackendPluginServer(BackendPluginServiceServicer):
+    def CreateTask(self, request: plugin_system_pb2.TaskCreateRequest, context) -> plugin_system_pb2.TaskCreateResponse:
+        req = TaskCreateRequest.from_flyte_idl(request)
+        plugin = BackendPluginRegistry.get_plugin(req.template.type)
+        return plugin.create(req.inputs, req.output_prefix, req.template)
+
+    def GetTask(self, request: plugin_system_pb2.TaskGetRequest, context):
+        plugin = BackendPluginRegistry.get_plugin(request.task_type)
+        return plugin.get(job_id=request.job_id, output_prefix=request.output_prefix, prev_state=request.prev_state)
+
+    def DeleteTask(self, request: plugin_system_pb2.TaskDeleteRequest, context):
+        plugin = BackendPluginRegistry.get_plugin(request.task_type)
+        return plugin.delete(request.job_id)
+```
+# Register a New plugin
+
+
+### Run a Flytekit Backend Plugin
+The workflow code running Snowflake, Databricks should not be changed. We can add a new config `enable_backend_flytekit_plugin_system`. If it's true, the job will be handled by plugin system. If can't find the plugin in the registry, falling back to use the default web api plugin in the propeller.
+
+### Secrets Mmanagement
+Secrets management should not be imposed using Flyte’s convention, though we should provide a simplified secrets reader using flytekits secret system?
+
+## Backend Plugin System Deployment
+There are two options to deploy backend plugin system. 
+
+### 1. All the plugin running in one deployment
+Pros: 
+* Easy to maintain.
+* One image, and one deployment yaml file.
+
+Cons: 
+* Dependecy conflict between two plugins.
+* Need to restart the deployment and rebuild the image when we add a new backend plugin to flytekit.
+
+### 2. One plugin per deployment
+
+Pros:
+* We can scale up the specific plugin deployments when the reqeusts increase in some plugins.
+* Only need to deploy the plugins that people will use.
+
+Cons:
+* Hard to maintain. (tons of images, and deployment yaml files)
```

Comment:
If all plugins need to be in one image/deployment, how would it work if users want to create plugins private to their organisation? At least for that use case a separate image/deployment would be nice. Otherwise they would have to fork flytekit to build the image?

Reply:
For plugins internal to an organization, I think we can have the mapping of plugin --> plugin address in the flyte configuration and let the plugin developer (or the system administrator) deploy the plugin separately from flyte.
e.g.

```
plugins:
    plugin-1: plugin-1.plugin-namespace.svc.cluster.local
    plugin-2: plugin-2.plugin-namespace.svc.cluster.local
```